File: libs/go/faults/code.go
// Copyright © 2026 Mindclade, LLC. All Rights Reserved.
// Mindclade Proprietary and Confidential.
// SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
//

package faults

import (
	"fmt"
	"strings"
)

// Code is a stable, transport-neutral classification for a failure.
//
// Codes are intentionally broad. Domain-specific machine-readable detail
// belongs in Fault.Reason rather than in an ever-growing code taxonomy.
//
// The canonical set mirrors mindclade.common.v1.ErrorCode in
// protocols/proto/mindclade/common/v1/errors.proto, which is the wire
// authority. tests/integration/cross_language/test_error_codes.py fails if this
// file, that proto, and libs/rust/faults drift apart.
type Code string

const (
	CodeUnknown            Code = "unknown"
	CodeCanceled           Code = "canceled"
	CodeInvalidArgument    Code = "invalid_argument"
	CodeDeadlineExceeded   Code = "deadline_exceeded"
	CodeNotFound           Code = "not_found"
	CodeAlreadyExists      Code = "already_exists"
	CodePermissionDenied   Code = "permission_denied"
	CodeUnauthenticated    Code = "unauthenticated"
	CodeResourceExhausted  Code = "resource_exhausted"
	CodeFailedPrecondition Code = "failed_precondition"
	CodeConflict           Code = "conflict"
	CodeAborted            Code = "aborted"
	CodeOutOfRange         Code = "out_of_range"
	CodeNotImplemented     Code = "not_implemented"
	CodeInternal           Code = "internal"
	CodeUnavailable        Code = "unavailable"
	CodeDataLoss           Code = "data_loss"
)

// canonicalCodes is the taxonomy itself, ordered by the tag its counterpart
// carries in mindclade.common.v1.ErrorCode. The order is part of what Codes
// documents, so the two files can be read side by side and a reordering shows
// up as a reviewable diff instead of disappearing into map iteration.
//
// This slice, not the const block above, is what makes a code canonical.
// validCodes used to be a second hand-maintained literal naming the same
// seventeen codes, so declaring a constant and forgetting its map entry left a
// code that every constructor silently rewrote to CodeUnknown — a spelling the
// compiler accepted and no test could see. Deriving the set removes that second
// place to be wrong.
var canonicalCodes = []Code{
	CodeInvalidArgument,
	CodeUnauthenticated,
	CodePermissionDenied,
	CodeNotFound,
	CodeConflict,
	CodeResourceExhausted,
	CodeDeadlineExceeded,
	CodeUnavailable,
	CodeInternal,
	CodeCanceled,
	CodeAlreadyExists,
	CodeFailedPrecondition,
	CodeAborted,
	CodeOutOfRange,
	CodeNotImplemented,
	CodeDataLoss,
	CodeUnknown,
}

var validCodes = func() map[Code]struct{} {
	set := make(map[Code]struct{}, len(canonicalCodes))
	for _, code := range canonicalCodes {
		set[code] = struct{}{}
	}
	return set
}()

// Codes returns the canonical taxonomy in the tag order of
// mindclade.common.v1.ErrorCode, excluding that enum's UNSPECIFIED value, which
// this package represents as the absence of a code rather than as a constant.
//
// It exists so the taxonomy can be compared against the generated protobuf
// descriptor by a program rather than by a regular expression over this file:
// protocols/consumers/_bazel/faults_conformance_test.go is that comparison, and
// it can only enumerate what this package exports. The returned slice is a
// fresh copy, so a caller cannot shorten the set a conformance check is about
// to read.
func Codes() []Code {
	return append([]Code(nil), canonicalCodes...)
}

// codeAliases are non-canonical spellings that must still resolve on ingestion.
//
// libs/rust/faults emitted these two spellings before the taxonomy was
// reconciled, so they are already present in peer responses, telemetry
// attributes, and stored diagnostics. Accepting them costs nothing; rejecting
// them would silently reclassify a canceled or unimplemented operation as
// CodeUnknown and discard the retry behavior the sender chose. They are
// deliberately not Valid: this package accepts both spellings and emits one.
var codeAliases = map[string]Code{
	"cancelled":     CodeCanceled,
	"unimplemented": CodeNotImplemented,
}

// CodeAliases returns the accepted non-canonical spellings and the code each
// resolves to, as a fresh map.
//
// Exported for the same reason as Codes: a conformance check has to be able to
// assert that no alias has become a spelling the wire authority also declares.
// If that ever happened, one concept would parse to two different codes
// depending on which side wrote the message, which is the exact failure the
// canceled/cancelled split already caused once.
func CodeAliases() map[string]Code {
	aliases := make(map[string]Code, len(codeAliases))
	for spelling, code := range codeAliases {
		aliases[spelling] = code
	}
	return aliases
}

// String returns the wire representation of the code.
func (c Code) String() string {
	return string(c)
}

// Valid reports whether c is one of the package's canonical codes.
func (c Code) Valid() bool {
	_, ok := validCodes[c]
	return ok
}

// ParseCode parses a canonical code or a recognized legacy spelling. Leading
// and trailing whitespace and ASCII letter case are normalized. Unrecognized
// values return CodeUnknown and an error.
func ParseCode(value string) (Code, error) {
	normalized := strings.ToLower(strings.TrimSpace(value))
	if code := Code(normalized); code.Valid() {
		return code, nil
	}
	if code, ok := codeAliases[normalized]; ok {
		return code, nil
	}
	return CodeUnknown, fmt.Errorf("faults: invalid code %q", value)
}

// NormalizeCode returns c when it is valid and CodeUnknown otherwise.
func NormalizeCode(c Code) Code {
	if c.Valid() {
		return c
	}
	return CodeUnknown
}

func defaultMessage(code Code) string {
	switch NormalizeCode(code) {
	case CodeCanceled:
		return "request canceled"
	case CodeInvalidArgument:
		return "invalid argument"
	case CodeDeadlineExceeded:
		return "deadline exceeded"
	case CodeNotFound:
		return "resource not found"
	case CodeAlreadyExists:
		return "resource already exists"
	case CodePermissionDenied:
		return "permission denied"
	case CodeUnauthenticated:
		return "authentication required"
	case CodeResourceExhausted:
		return "resource exhausted"
	case CodeFailedPrecondition:
		return "failed precondition"
	case CodeConflict:
		return "resource conflict"
	case CodeAborted:
		return "operation aborted"
	case CodeOutOfRange:
		return "value out of range"
	case CodeNotImplemented:
		return "operation not implemented"
	case CodeInternal:
		return "internal error"
	case CodeUnavailable:
		return "service unavailable"
	case CodeDataLoss:
		return "data loss"
	case CodeUnknown:
		fallthrough
	default:
		return "operation failed"
	}
}

File: libs/rust/faults/src/code.rs
// Copyright © 2026 Mindclade, LLC. All Rights Reserved.
// Mindclade Proprietary and Confidential.
// SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
//

//! Stable machine-readable error codes.
//!
//! The set mirrors `mindclade.common.v1.ErrorCode` in
//! `protocols/proto/mindclade/common/v1/errors.proto`, which is the wire
//! authority, and `libs/go/faults`, which is the other mirror.
//! `tests/integration/cross_language/test_error_codes.py` fails if the three
//! drift apart.

use core::fmt;
use core::str::FromStr;

/// Longest canonical spelling is `failed_precondition` at 19 bytes. Parsing is
/// an ingestion path fed by peers, so it rejects on length before it allocates
/// a normalized copy rather than lowercasing an arbitrarily long peer string.
const MAXIMUM_CODE_BYTES: usize = 32;

/// Stable classes used across process and protocol boundaries.
///
/// Deliberately **not** `#[non_exhaustive]`. Every crate in this workspace is
/// `publish = false`, so the attribute bought no external-API stability; what
/// it bought instead was a compulsory `_ =>` arm in every downstream `match`,
/// and four transport edges used that arm to render `NotFound`, `Aborted`,
/// `Unimplemented`, and `DataLoss` as HTTP 500 / gRPC `internal`. Exhaustive
/// matching is the mechanism that keeps a code added here from silently
/// inheriting a catch-all at an edge that has not been taught about it.
///
/// Totality on ingestion — the reason a taxonomy usually reaches for the
/// attribute — is provided at the *value* level instead, by [`Code::Unknown`]
/// and [`Code::from_wire`]: a peer's unrecognized code still parses, it just
/// parses into a code every match already names.
#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub enum Code {
    /// Classification failed, or a peer sent a code this build does not know.
    ///
    /// This is the landing spot that keeps ingestion total. Without it a peer's
    /// own catch-all — `libs/go/faults` emits `"unknown"` exactly when it cannot
    /// classify — was a code this crate could not parse at all.
    Unknown,
    InvalidArgument,
    NotFound,
    AlreadyExists,
    FailedPrecondition,
    Aborted,
    OutOfRange,
    Unimplemented,
    Internal,
    Unavailable,
    DataLoss,
    Unauthenticated,
    PermissionDenied,
    ResourceExhausted,
    DeadlineExceeded,
    Cancelled,
    Conflict,
}

impl Code {
    /// Returns the stable snake-case wire representation.
    ///
    /// `Cancelled` and `Unimplemented` serialize as `canceled` and
    /// `not_implemented`: those are the spellings already on
    /// `Mindclade-Error-Code` headers and in telemetry attributes emitted by the
    /// Go control plane, so they are the canonical ones. The variant names stay
    /// as they are because they mirror the tonic/gRPC names used at this
    /// crate's own transport boundary.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Unknown => "unknown",
            Self::InvalidArgument => "invalid_argument",
            Self::NotFound => "not_found",
            Self::AlreadyExists => "already_exists",
            Self::FailedPrecondition => "failed_precondition",
            Self::Aborted => "aborted",
            Self::OutOfRange => "out_of_range",
            Self::Unimplemented => "not_implemented",
            Self::Internal => "internal",
            Self::Unavailable => "unavailable",
            Self::DataLoss => "data_loss",
            Self::Unauthenticated => "unauthenticated",
            Self::PermissionDenied => "permission_denied",
            Self::ResourceExhausted => "resource_exhausted",
            Self::DeadlineExceeded => "deadline_exceeded",
            Self::Cancelled => "canceled",
            Self::Conflict => "conflict",
        }
    }
    /// Parses a wire code, degrading anything unrecognized to [`Code::Unknown`].
    ///
    /// Use this wherever the input comes from a peer. [`FromStr`] stays strict
    /// so configuration and test vectors still fail loudly on a typo, but a
    /// remote fault must never be rejected because the sender runs a build that
    /// knows one more code than this one does.
    #[must_use]
    pub fn from_wire(value: &str) -> Self {
        Self::from_str(value).unwrap_or(Self::Unknown)
    }
    /// Whether failures with this code are commonly transient.
    #[must_use]
    pub const fn is_transient(self) -> bool {
        matches!(
            self,
            Self::Aborted
                | Self::Unavailable
                | Self::ResourceExhausted
                | Self::DeadlineExceeded
                | Self::Conflict
        )
    }
}

impl fmt::Display for Code {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(self.as_str())
    }
}

/// Returned when an unknown wire code is parsed.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ParseCodeError(String);

impl fmt::Display for ParseCodeError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "unknown error code: {}", self.0)
    }
}

impl std::error::Error for ParseCodeError {}

impl FromStr for Code {
    type Err = ParseCodeError;
    fn from_str(value: &str) -> Result<Self, Self::Err> {
        let trimmed = value.trim();
        if trimmed.len() > MAXIMUM_CODE_BYTES {
            // Echo a bounded prefix; the rejected value came from a peer and the
            // error text reaches logs.
            let mut end = MAXIMUM_CODE_BYTES;
            while end > 0 && !trimmed.is_char_boundary(end) {
                end -= 1;
            }
            return Err(ParseCodeError(trimmed[..end].to_owned()));
        }
        // Case is normalized to match `faults.ParseCode` on the Go side, so the
        // same input resolves the same way in both mirrors.
        let normalized = trimmed.to_ascii_lowercase();
        // The second spelling in each pair is the one this crate emitted before
        // the taxonomy was reconciled. Both mirrors accept both spellings and
        // emit one, so values already in flight keep resolving.
        let code = match normalized.as_str() {
            "unknown" => Self::Unknown,
            "invalid_argument" => Self::InvalidArgument,
            "not_found" => Self::NotFound,
            "already_exists" => Self::AlreadyExists,
            "failed_precondition" => Self::FailedPrecondition,
            "aborted" => Self::Aborted,
            "out_of_range" => Self::OutOfRange,
            "not_implemented" | "unimplemented" => Self::Unimplemented,
            "internal" => Self::Internal,
            "unavailable" => Self::Unavailable,
            "data_loss" => Self::DataLoss,
            "unauthenticated" => Self::Unauthenticated,
            "permission_denied" => Self::PermissionDenied,
            "resource_exhausted" => Self::ResourceExhausted,
            "deadline_exceeded" => Self::DeadlineExceeded,
            "canceled" | "cancelled" => Self::Cancelled,
            "conflict" => Self::Conflict,
            _ => return Err(ParseCodeError(normalized)),
        };
        Ok(code)
    }
}
